### ISSUE TITLE
Updating whole data frames in recursive lists of data frames with classes = "data.frame" and how = "recurse" not working

### ISSUE THREAD

**Issue by hermidalc** [JorisChau/rrapply#5]
`rrapply` is a wonderful library and really very useful for work with `GenomicDataCommons` query results data structures, which can be a deeply nested list of recursive lists of data frames. I'm finding a possible bug or really could be something I'm missing.

If you run this query and retrieve the results:

```R
library(GenomicDataCommons)
library(rrapply)

stopifnot(GenomicDataCommons::status()$status == "OK")
gdc_query <-
    files() %>%
    GenomicDataCommons::filter(
        cases.project.project_id == "TCGA-BRCA"
        & cases.samples.sample_type == "Primary Tumor"
        & analysis.workflow_type == "HTSeq - Counts"
    ) %>%
    GenomicDataCommons::select(c(
        "file_name",
        "analysis.workflow_type",
        "cases.project.project_id",
        "cases.case_id",
        "cases.submitter_id",
        "cases.samples.sample_id",
        "cases.samples.submitter_id",
        "cases.samples.sample_type",
        "cases.samples.is_ffpe",
        "cases.samples.portions.portion_id",
        "cases.samples.portions.submitter_id",
        "cases.samples.portions.is_ffpe",
        "cases.samples.portions.analytes.aliquots.aliquot_id",
        "cases.samples.portions.analytes.aliquots.submitter_id"
    ))
gdc_results <- results_all(gdc_query)
```
Example the `portions` level by:

```R
print(sapply(sapply(gdc_results$cases, `[[`, "samples"), `[[`, "portions"))
```

You will see at there are a few nested data frames that have `NULL` entries:

```
$`58d983ee-8d34-4276-95d2-a660a87f1c2e`
                            portion_id                                                           analytes        submitter_id    state is_ffpe
1 78b36e1c-89ca-40dd-b0f7-ccb3de92d60d de01d9b6-e29d-495e-815a-c7d2a7c15af3, TCGA-PL-A8LZ-01A-31R-A36F-07 TCGA-PL-A8LZ-01A-31 released   FALSE
2 3cd08852-4369-5963-b564-fab2959b6691                                                               NULL                <NA>     <NA>      NA
3 bbd3b2b8-afe1-5739-9a57-0b095e2ab66c                                                               NULL                <NA>     <NA>      NA

$`2154c9df-56e8-4281-b917-9618ea1224dc`
                            portion_id                                                           analytes        submitter_id    state is_ffpe
1 b549dda5-678d-49cd-b464-609717682062 2ca5942e-dcbf-46f1-8427-1ea7216dc559, TCGA-PL-A8LY-01A-11R-A41B-07 TCGA-PL-A8LY-01A-11 released   FALSE
2 1dd8e696-aecd-54f2-88d6-ea2c9376b219                                                               NULL                <NA>     <NA>      NA
```

I first fix the `NULL` with `NA` using `rrapply` which works:

```R
gdc_results$cases <- rrapply(
    gdc_results$cases, f=function(x) NA, condition = is.null, how="replace"
)
```

But then to fix these whole data frames by removing rows which have a nested `analytes` list entry that is `NA` `rrapply` doesn't work here:

```R
gdc_results$cases <- rrapply(
    gdc_results$cases,
    f=function(x) {
        if ("analytes" %in% colnames(x))
            x[!is.na(unlist(x$analytes, recursive=FALSE)), , drop=FALSE]
        else
            x
    },
    classes="data.frame", how="recurse"
)
```

What am I doing wrong?

**Comment by hermidalc** [comment]
If I unpack things down to the portion level then it will work:

```R
rrapply(
    sapply(sapply(gdc_results$cases, `[[`, "samples"), `[[`, "portions"), f=function(x) {
    if ("analytes" %in% colnames(x))
        x[!is.na(unlist(x$analytes, recursive=FALSE)), , drop=FALSE]
    else
        x
    }, classes="data.frame", how="recurse" 
) 
```

But not sure why it's not recursing down from the case or even the sample level to alter these while data frames, even though I am using `how = "recurse"`

**Comment by hermidalc** [comment]
Sorry now I see in the docs:

`how = "recurse" is a specialized option that is only useful in combination with e.g.
classes = "list" to recurse further into updated “list-like” elements. This is explained in more
detail below.`

So maybe this would be a worthwhile new feature to add, the use case that one would want the function to be applied to the entire data frame, but then after applying it to still continuing recursing into any nested lists inside the data frame. Or if you have any other suggestions on how I could get this to work with the current functionality.

**Comment by JorisChau** [comment]
@hermidalc: `how = "recurse"` will recurse further into modified elements of the nested list, but I don't think that is what is needed here, i.e. recursing into an `analytes` column you will not find another (deeper) `analytes` column. It seems that the default `how = "replace"` should be sufficient to filter the nested data.frames:

```r
## data.frames with missing analytes
## nb: classes should also include "list" since some of the 
## higher-level nodes (e.g. samples and portions) are lists 
## and not data.frames and otherwise rrapply will skip these nodes
rrapply(
  gdc_results,
  classes = c("list", "data.frame"),
  condition = function(x) "analytes" %in% colnames(x) && length(Filter(is.null, x$analytes)) > 0,
  how = "flatten"
)
#> $`1`
#>                             portion_id                                                           analytes        submitter_id is_ffpe
#> 1 e4a7a7e7-7a66-4440-93d3-f5bf0bc86d80 34f3d207-0bfe-4699-b01b-d145942ba488, TCGA-PL-A8LV-01A-21R-A41B-07 TCGA-PL-A8LV-01A-21   FALSE
#> 2 36cfb9ad-1b54-5ba1-873b-7ca07865b539                                                               NULL                <NA>      NA
#> 
#> $`1`
#>                             portion_id                                                           analytes        submitter_id is_ffpe
#> 1 f07ce95f-eaeb-4dbd-b0a3-4f3c406b3044 19cadb60-b213-4fd3-b5ad-d2b49000b6a3, TCGA-PL-A8LX-01A-11R-A41B-07 TCGA-PL-A8LX-01A-11   FALSE
#> 2 09abe53f-204c-51e0-9a51-9a48ee0ce7ab                                                               NULL                <NA>      NA
#> 
#> $`1`
#>                             portion_id                                                           analytes        submitter_id is_ffpe
#> 1 78b36e1c-89ca-40dd-b0f7-ccb3de92d60d de01d9b6-e29d-495e-815a-c7d2a7c15af3, TCGA-PL-A8LZ-01A-31R-A36F-07 TCGA-PL-A8LZ-01A-31   FALSE
#> 2 3cd08852-4369-5963-b564-fab2959b6691                                                               NULL                <NA>      NA
#> 3 bbd3b2b8-afe1-5739-9a57-0b095e2ab66c                                                               NULL                <NA>      NA
#> 
#> $`1`
#>                             portion_id                                                           analytes        submitter_id is_ffpe
#> 1 b549dda5-678d-49cd-b464-609717682062 2ca5942e-dcbf-46f1-8427-1ea7216dc559, TCGA-PL-A8LY-01A-11R-A41B-07 TCGA-PL-A8LY-01A-11   FALSE
#> 2 1dd8e696-aecd-54f2-88d6-ea2c9376b219                                                               NULL                <NA>      NA

## remove rows with missing analytes
gdc_results1 <- rrapply(
  gdc_results,
  classes = c("list", "data.frame"),
  condition = function(x) "analytes" %in% colnames(x) && length(Filter(is.null, x$analytes)) > 0,
  f = function(x) x[sapply(x$analytes, Negate(is.null)), ]
)

## check again for data.frames with missing analytes
rrapply(
  gdc_results1,
  classes = c("list", "data.frame"),
  condition = function(x) "analytes" %in% colnames(x) && length(Filter(is.null, x$analytes)) > 0,
  how = "flatten"
)
#> named list()
 
## filter updated portions data.frame
rrapply(
  gdc_results1,
  classes = c("list", "data.frame"),
  condition = function(x) "portion_id" %in% colnames(x) && "e4a7a7e7-7a66-4440-93d3-f5bf0bc86d80" %in% x$portion_id,
  how = "flatten"
)
#> $`1`
#>                             portion_id                                                           analytes        submitter_id is_ffpe
#> 1 e4a7a7e7-7a66-4440-93d3-f5bf0bc86d80 34f3d207-0bfe-4699-b01b-d145942ba488, TCGA-PL-A8LV-01A-21R-A41B-07 TCGA-PL-A8LV-01A-21   FALSE
```

**Comment by JorisChau** [comment]
@hermidalc: I don't understand your previous question exactly, but having another look at the stated issue there indeed is a problem when combining `classes = "data.frame"` with `how = "recurse"`. As a simpler illustration, this call should convert all nested data.frames to sublists:

```r
rrapply(gdc_results, classes = "data.frame", f = as.list, how = "recurse")
```
but only replaces the top-level data.frames, (as is expected with `rrapply(gdc_results, classes = "data.frame", f = as.list, how = "replace")`).

The following call *does* convert all nested data.frames to lists recursively, but the same result should be obtained with `classes = "data.frame"` only:

```r
rrapply(gdc_results, classes = c("list", "data.frame"), f = as.list, how = "recurse")
```

**Comment by hermidalc** [comment]
Ok see this nested data frame from your post above:
```
                            portion_id                                                           analytes        submitter_id is_ffpe
1 78b36e1c-89ca-40dd-b0f7-ccb3de92d60d de01d9b6-e29d-495e-815a-c7d2a7c15af3, TCGA-PL-A8LZ-01A-31R-A36F-07 TCGA-PL-A8LZ-01A-31   FALSE
2 3cd08852-4369-5963-b564-fab2959b6691                                                               NULL                <NA>      NA
3 bbd3b2b8-afe1-5739-9a57-0b095e2ab66c                                                               NULL                <NA>      NA
```

Will your code here below continue recursing down the nested list inside the data.frame `analytes` column first row entry while also updating the data.frame by removing the second two rows which have `NULL`s?

```R
gdc_results1 <- rrapply(
  gdc_results,
  classes = c("list", "data.frame"),
  condition = function(x) "analytes" %in% colnames(x) && length(Filter(is.null, x$analytes)) > 0,
  f = function(x) x[!sapply(x$analytes, is.null), ]
)
```

**Comment by JorisChau** [comment]
@hermidalc: No it will not (using `how = "replace"`), it replaces the data.frame element by the modified data.frame after application of `f` and not recurse further into the modified data.frame. `how = "recurse"` is meant for this purpose, but as you pointed out, it is currently not behaving as expected in combination with `classes = "data.frame"`. 

NB: I think my confusion is due to the fact that using `how = "recurse"` is not needed in this example (as recursing into an `analytes` column, you will not find another data.frame satisfying the same condition).

**Comment by hermidalc** [comment]
> NB: I think my confusion is due to the fact that using `how = "recurse"` is not needed in this example (as recursing into an `analytes` column, you will not find another data.frame satisfying the same condition).

In this example no, but in the more general use case yes. Imagine that at each recursion level where you have a list of data.frames, each data.frame can have one or more columns where each column is a list (of the same length as the number of data.frame rows). Each list can have some `NULL` entries, which mean those data.frame rows need to be removed, but others will have valid data.frames, and in each of those data.frames the recursion starts over again with the same possibilities as above until we reach the leaves of the data structure.

Can `rrapply` process this, even with multiple calls to it? Like with each call altering things to make the next pass logic possible.

**Comment by JorisChau** [comment]
Yes it should be able to with `how = "recurse"`, as also mentioned in the previous comment. Here is a simpler example to illustrate:

```r
## nested data.frames with list columns
ex <- list(data.frame(v1 = c(1, 2), v2 = I(list(data.frame(v1 = c(1, 2), v2 = I(list(1, NULL))), NULL))))

str(ex)
#> List of 1
#>  $ :'data.frame':    2 obs. of  2 variables:
#>   ..$ v1: num [1:2] 1 2
#>   ..$ v2:List of 2
#>   .. ..$ :'data.frame':  2 obs. of  2 variables:
#>   .. .. ..$ v1: num [1:2] 1 2
#>   .. .. ..$ v2:List of 2
#>   .. .. .. ..$ : num 1
#>   .. .. .. ..$ : NULL
#>   .. .. .. ..- attr(*, "class")= chr "AsIs"
#>   .. ..$ : NULL
#>   .. ..- attr(*, "class")= chr "AsIs"

## cleaned nested data.frames
ex1 <- rrapply(
  ex, 
  classes = c("list", "data.frame"),
  f = function(x) x[!sapply(x$v2, is.null), ],
  how = "recurse"
)

str(ex1)
#> List of 1
#>  $ :'data.frame':    1 obs. of  2 variables:
#>   ..$ v1: num 1
#>   ..$ v2:List of 1
#>   .. ..$ :'data.frame':  1 obs. of  2 variables:
#>   .. .. ..$ v1: num 1
#>   .. .. ..$ v2:List of 1
#>   .. .. .. ..$ : num 1
#>   .. .. .. ..- attr(*, "class")= chr "AsIs"
#>   .. ..- attr(*, "class")= chr "AsIs"
```

**Comment by hermidalc** [comment]
> ## cleaned nested data.frames
> ex1 <- rrapply(
>   ex, 
>   classes = c("list", "data.frame"),
>   f = function(x) x[!sapply(x$v2, is.null), ],
>   how = "recurse"
> )

Sorry I must be totally missing something, when I run this on the OP query results data structure:
```R
rrapply(
  gdc_results$cases,
  classes = c("list", "data.frame"),
  f = function(x) x[!sapply(x$analytes, is.null), ],
  how = "recurse"
)
```
I get
```
Error in x[!sapply(x$analytes, is.null), ] : 
  incorrect number of dimensions
```

**Comment by JorisChau** [comment]
@hermidalc Perhaps you already solved it by now, but the `condition` argument is missing. The above call applies `f` to every list/data.frame element in the nested list (instead of only data.frames with an `analytes` column), which also explains the error.

**Comment by hermidalc** [comment]
Yep I realized it eventually, thank you for all your help @JorisChau. As you can see `rrapply` is extremely useful to post-process `GenomicDataCommons` query API data structures (R representation of JSON), so I’m very glad you wrote it

**Comment by JorisChau** [comment]
This is fixed in cran-release v1.2.4 (commit a116683cf6eb643469bddbed26273f3b0c55816e)
